rados/read_op_exec.go

```diff
@@ -0,0 +1,114 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <stdlib.h>
+// #include <rados/librados.h>
+//
+import "C"
+
+import (
+	"runtime"
+	"unsafe"
+)
+
+// ReadOpExecStep - step for exec operation code.
+type ReadOpExecStep struct {
+	charPtr *C.char
+	size    C.size_t
+	prval   *C.int
+}
+
+// newExecStepOp - init new *execStepOp.
+func newReadOpExecStep() *ReadOpExecStep {
+	es := &ReadOpExecStep{
+		charPtr: nil,
+		size:    0,
+		prval:   (*C.int)(C.malloc(C.sizeof_int)),
+	}
+	*es.prval = 0
+
+	runtime.SetFinalizer(es, func(es *ReadOpExecStep) {
+		if es != nil {
+			es.free()
+			es.freeBuffer()
+			es = nil
+		}
+	})
+	return es
+}
+
+// free - will be called to free any resources, especially C memory.
+func (es *ReadOpExecStep) free() {
+	if es.prval != nil {
+		C.free(unsafe.Pointer(es.prval))
+		es.prval = nil
+	}
+}
+
+// freeBuffer - releases C allocated buffer. It is separated from es.free() because lifespan of C allocated buffer is
+// longer than lifespan of read operation.
+func (es *ReadOpExecStep) freeBuffer() {
+	if es.charPtr != nil {
+		C.free(unsafe.Pointer(es.charPtr))
+		es.charPtr = nil
+	}
+}
+
+// update - update state operation.
+func (es *ReadOpExecStep) update() error {
+	return getError(*es.prval)
+}
+
+// Bytes - go slice of bytes allocated in C.
+func (es *ReadOpExecStep) Bytes() []byte {
+	return C.GoBytes(unsafe.Pointer(es.charPtr), C.int(es.size))
+}
+
+// Exec - execute an OSD class method on an object See rados_exec() for general description.
+// The output buffer is allocated on the heap; the caller is expected to release that memory with
+// rados_buffer_free(). The buffer and length pointers can all be NULL, in which case they are not filled in.
+//
+// Implements:
+//
+//	void rados_read_op_exec(rados_read_op_t read_op,
+//	                        const char *cls,
+//	                        const char *method,
+//	                        const char *in_buf,
+//	                        size_t in_len,
+//	                        char **out_buf,
+//	                        size_t *out_len,
+//	                        int *prval);
+func (r *ReadOp) Exec(in []byte, clsName, method string) (*ReadOpExecStep, error) {
+	cClsName := C.CString(clsName)
+	defer C.free(unsafe.Pointer(cClsName))
+
+	cMethod := C.CString(method)
+	defer C.free(unsafe.Pointer(cMethod))
+
+	var inBuff *C.char
+	var inBuffLen C.size_t
+
+	if len(in) > 0 {
+		rs := newReadStep(in, 0)
+		r.steps = append(r.steps, rs)
+		inBuff = rs.cBuffer
+		inBuffLen = rs.cReadLen
```

Comment:
I took a 2nd look at the C documentation of rados_read_op_exec and am now a little confused. Aren't the "in" buffer and length for passing input to the custom method? How are users of the api supposed to pass a string/buffer into the exec method?

Reply:
This was done because if user passes nil slice it will fail [here](https://github.com/ceph/go-ceph/blob/master/rados/read_step.go#L27). Also in that case we do not need read step

Reply:
I now see my source of some of my confusion. The `in` argument in `Exec` in the Go codes is the first argument, before the class name and method. However, in the C api the `inBuff` and `inBuffLen` args come *after* the class name and method. Can you please reorder the arguments to be `Exec(clsName, method string, in []byte)` - this is both more consistent with the C api and a more natural flow (IMO).

Furthermore, I think you can combine the `in` values into the single `ReadOpExecStep` and avoid creating two exec steps for this single call.